astropy/nddata/nddata.py

```diff
@@ -186,30 +187,70 @@ def __init__(
             elif data.meta is not None:
                 meta = data.meta
 
+            # get the data attribute as it is, and continue to process it:
             data = data.data
 
-        else:
-            if hasattr(data, "mask") and hasattr(data, "data"):
-                # Separating data and mask
+        # if the data is wrapped by astropy.utils.masked.Masked:
+        if isinstance(data, Masked):
+            # first get the mask if one is available:
+            if hasattr(data, "mask"):
                 if mask is not None:
                     log.info(
-                        "overwriting Masked Objects's current mask with specified mask."
+                        "overwriting Masked Quantity's current mask with specified mask."
                     )
                 else:
                     mask = data.mask
 
-                # Just save the data for further processing, we could be given
-                # a masked Quantity or something else entirely. Better to check
-                # it first.
-                data = data.data
+            if isinstance(data, MaskedNDArray):
```

Comment:
I assume that here and below where you checked for class type it is because checking for the presence of a `mask` attribute won't let you distinguish between the two kinds of masked things that might be passed in?

Reply:
Exactly.